File: e1_precheck.js
const { chromium } = require('playwright');
const fs = require('fs');
const path = require('path');

(async () => {
  const urls = [
    'https://lying-www.hubbuyer.com',
    'https://lying-www.hubbuyer.com/fee',
    'https://lying-b2b.hubbuyer.com',
    'https://lying-admin.hubbuyer.com'
  ];
  const outDir = path.join(process.cwd(), 'output', 'playwright', 'agent2_precheck');
  fs.mkdirSync(outDir, { recursive: true });
  const browser = await chromium.launch({ headless: true });
  const context = await browser.newContext({ viewport: { width: 1440, height: 1000 }, ignoreHTTPSErrors: true });
  const page = await context.newPage();
  const results = [];
  for (const url of urls) {
    let status = null, title = '', finalUrl = '', error = '';
    try {
      const resp = await page.goto(url, { waitUntil: 'domcontentloaded', timeout: 45000 });
      status = resp ? resp.status() : null;
      await page.waitForLoadState('networkidle', { timeout: 15000 }).catch(()=>{});
      title = await page.title();
      finalUrl = page.url();
    } catch (e) {
      error = e.message;
      finalUrl = page.url();
      title = await page.title().catch(()=> '');
    }
    results.push({ url, status, title, finalUrl, error });
  }

  const admin = { loginAttempted: false, loginSuccess: false, reason: '', screenshot: '', finalUrl: '', title: '', cookies: 0, storageKeys: [] };
  try {
    admin.loginAttempted = true;
    const resp = await page.goto('https://lying-admin.hubbuyer.com', { waitUntil: 'domcontentloaded', timeout: 45000 });
    await page.waitForLoadState('networkidle', { timeout: 15000 }).catch(()=>{});
    admin.title = await page.title();
    // Identify likely login controls.
    const info = await page.evaluate(() => {
      const inputs = [...document.querySelectorAll('input')].map((el, i) => ({
        i,
        type: el.type || '',
        name: el.getAttribute('name') || '',
        id: el.id || '',
        placeholder: el.getAttribute('placeholder') || '',
        autocomplete: el.getAttribute('autocomplete') || '',
        visible: !!(el.offsetWidth || el.offsetHeight || el.getClientRects().length)
      }));
      const buttons = [...document.querySelectorAll('button')].map((el, i) => ({i, text: (el.innerText||el.textContent||'').trim(), type: el.type || '', visible: !!(el.offsetWidth || el.offsetHeight || el.getClientRects().length)}));
      return {inputs, buttons, bodyText: document.body.innerText.slice(0,1000)};
    });

    const userSelectors = [
      'input[name="username"]','input[name="userName"]','input[name="account"]','input[name="email"]','input[id*="user" i]','input[placeholder*="账号"]','input[placeholder*="用户"]','input[placeholder*="邮箱"]','input[type="text"]'
    ];
    const passSelectors = ['input[type="password"]','input[name="password"]','input[placeholder*="密码"]'];
    async function firstVisible(selectors) {
      for (const s of selectors) {
        const loc = page.locator(s).filter({ hasNotText: '' }).first();
        try { if (await loc.count() && await loc.isVisible({timeout:1000})) return loc; } catch(e) {}
      }
      return null;
    }
    const user = await firstVisible(userSelectors);
    const pass = await firstVisible(passSelectors);
    if (!user || !pass) {
      admin.reason = '未找到可见账号/密码输入框；页面控件信息：' + JSON.stringify(info).slice(0,2000);
      const p = path.join(outDir, 'env_login_page.webp');
      await page.screenshot({ path: p, fullPage: true, type: 'webp', quality: 80 });
      admin.screenshot = p;
    } else {
      await user.click(); await user.fill('admin');
      await pass.click(); await pass.fill('123333');
      const before = page.url();
      const loginText = /登录|登 录|Login|Sign in|提交|确定/;
      let clicked = false;
      const btns = page.getByRole('button', { name: loginText });
      if (await btns.count()) { await btns.first().click(); clicked = true; }
      else {
        const submit = page.locator('button[type="submit"], input[type="submit"]').first();
        if (await submit.count()) { await submit.click(); clicked = true; }
        else { await pass.press('Enter'); clicked = true; }
      }
      await page.waitForLoadState('networkidle', { timeout: 20000 }).catch(()=>{});
      await page.waitForTimeout(3000);
      admin.finalUrl = page.url();
      admin.title = await page.title();
      const cookies = await context.cookies();
      admin.cookies = cookies.length;
      admin.storageKeys = await page.evaluate(() => Object.keys(localStorage).concat(Object.keys(sessionStorage).map(k=>'session:'+k))).catch(()=>[]);
      const body = await page.locator('body').innerText({timeout:5000}).catch(()=> '');
      const hasLoginForm = await page.locator('input[type="password"]').count().catch(()=>0);
      const hasError = /密码错误|账号.*错误|用户名.*错误|验证码|captcha|失败|invalid|error|Incorrect/i.test(body);
      admin.loginSuccess = (admin.finalUrl !== before || hasLoginForm === 0) && !hasError && (cookies.length > 0 || admin.storageKeys.length > 0);
      if (!admin.loginSuccess) {
        admin.reason = `登录后仍未确认有效会话。finalUrl=${admin.finalUrl}; hasPasswordInputs=${hasLoginForm}; cookies=${cookies.length}; storageKeys=${admin.storageKeys.length}; pageText=${body.slice(0,800)}`;
        const p = path.join(outDir, 'env_login_page.webp');
        await page.screenshot({ path: p, fullPage: true, type: 'webp', quality: 80 });
        admin.screenshot = p;
      } else {
        const p = path.join(outDir, 'env_precheck.webp');
        await page.screenshot({ path: p, fullPage: true, type: 'webp', quality: 80 });
        admin.screenshot = p;
      }
    }
  } catch (e) {
    admin.reason = e.stack || e.message;
    try {
      const p = path.join(outDir, 'env_login_page.webp');
      await page.screenshot({ path: p, fullPage: true, type: 'webp', quality: 80 });
      admin.screenshot = p;
    } catch (_) {}
  }
  console.log(JSON.stringify({ results, admin }, null, 2));
  await browser.close();
})();

File: 测试项目/会员体系/测试文档/执行报告/scripts/e1_precheck.py
"""
E1 环境预检脚本 (Layer 2 - Playwright Headless)
检查项: URL可达性 / Admin登录 / B2B登录 / 截图
"""
from playwright.sync_api import sync_playwright
import os, sys, io, json

sys.stdout = io.TextIOWrapper(sys.stdout.buffer, encoding='utf-8', errors='replace')

SCREENSHOTS = r"d:\test_workspace\会员体系\测试文档\执行报告\screenshots"
os.makedirs(SCREENSHOTS, exist_ok=True)

TARGETS = {
    "admin": "https://lying-admin.hubbuyer.com",
    "b2b": "https://lying-b2b.hubbuyer.com",
    "www": "https://lying-www.hubbuyer.com",
}

ADMIN_CRED = {"user": "admin", "pass": "123333"}
B2B_CRED = {"user": "359901314@qq.com", "pass": "123456"}

results = {}

def main():
    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context(viewport={"width": 1280, "height": 720})
        page = context.new_page()

        # === Check 1: URL Reachability ===
        print("=" * 60)
        print("[E1-1] URL 可达性检查")
        print("=" * 60)
        for name, url in TARGETS.items():
            try:
                resp = page.goto(url, wait_until="domcontentloaded", timeout=15000)
                status = resp.status if resp else "N/A"
                title = page.title()
                ok = status == 200 or status == 301 or status == 302
                results[f"url_{name}"] = "Pass" if ok else "Fail"
                print(f"  [{name}] {url}")
                print(f"    Status: {status} | Title: {title} | {'✅ Pass' if ok else '❌ Fail'}")
                page.screenshot(path=os.path.join(SCREENSHOTS, f"env_url_{name}.webp"))
            except Exception as e:
                results[f"url_{name}"] = "Fail"
                print(f"  [{name}] {url}")
                print(f"    ❌ Fail: {e}")

        # === Check 2: Admin Login ===
        print("\n" + "=" * 60)
        print("[E1-2] Admin 登录验证")
        print("=" * 60)
        try:
            page.goto("https://lying-admin.hubbuyer.com/login", wait_until="networkidle", timeout=20000)
            page.wait_for_selector("input", timeout=10000)
            inputs = page.query_selector_all("input")
            
            # 填入凭证
            inputs[0].fill(ADMIN_CRED["user"])
            inputs[1].fill(ADMIN_CRED["pass"])
            
            # 点击登录
            btn = page.query_selector("button")
            if btn:
                btn.click()
            
            # 等待跳转
            page.wait_for_timeout(3000)
            page.wait_for_load_state("networkidle", timeout=15000)
            
            current_url = page.url
            is_login_page = "/login" in current_url
            
            if not is_login_page:
                results["admin_login"] = "Pass"
                print(f"  ✅ 登录成功! 跳转至: {current_url}")
                
                # 提取 Cookies 供后续使用
                cookies = context.cookies()
                cookie_str = "; ".join([f"{c['name']}={c['value']}" for c in cookies])
                print(f"  Cookies 数量: {len(cookies)}")
                
                # 保存 cookies 到文件供 Layer 1 API 使用
                cookie_file = os.path.join(SCREENSHOTS, "..", "scripts", "helpers", "admin_cookies.json")
                os.makedirs(os.path.dirname(cookie_file), exist_ok=True)
                with open(cookie_file, 'w') as f:
                    json.dump(cookies, f, ensure_ascii=False)
                print(f"  Cookies 已保存至 admin_cookies.json")
            else:
                results["admin_login"] = "Fail"
                print(f"  ❌ 登录失败，仍在登录页: {current_url}")
            
            page.screenshot(path=os.path.join(SCREENSHOTS, "env_admin_login.webp"))
        except Exception as e:
            results["admin_login"] = "Fail"
            print(f"  ❌ Admin 登录异常: {e}")
            page.screenshot(path=os.path.join(SCREENSHOTS, "env_admin_login_error.webp"))

        # === Check 3: 核心管理页面可达 ===
        print("\n" + "=" * 60)
        print("[E1-3] 核心管理页面可达性")
        print("=" * 60)
        admin_pages = {
            "会员等级基础库": "https://lying-admin.hubbuyer.com/b2b/member/level",
            "全球会员定价": "https://lying-admin.hubbuyer.com/b2b/member/country",
            "全球服务定价": "https://lying-admin.hubbuyer.com/b2b/member/service",
        }
        for name, url in admin_pages.items():
            try:
                page.goto(url, wait_until="networkidle", timeout=15000)
                page.wait_for_timeout(2000)
                current = page.url
                title = page.title()
                # 如果被重定向回登录页说明 session 丢失
                if "/login" in current:
                    results[f"page_{name}"] = "Fail"
                    print(f"  [{name}] ❌ Session 丢失，被重定向至登录页")
                else:
                    results[f"page_{name}"] = "Pass"
                    print(f"  [{name}] ✅ 可达 | URL: {current}")
                safe_name = name.replace("/", "_")
                page.screenshot(path=os.path.join(SCREENSHOTS, f"env_page_{safe_name}.webp"))
            except Exception as e:
                results[f"page_{name}"] = "Fail"
                print(f"  [{name}] ❌ 异常: {e}")

        browser.close()

    # === Gate Decision ===
    print("\n" + "=" * 60)
    print("[E1] GATE DECISION")
    print("=" * 60)
    all_pass = all(v == "Pass" for v in results.values())
    for k, v in results.items():
        symbol = "✅" if v == "Pass" else "❌"
        print(f"  {symbol} {k}: {v}")
    
    if all_pass:
        print("\n🟢 环境预检通过，进入队列编排")
    else:
        print("\n🔴 环境预检存在失败项，请查看详情")
    
    return 0 if all_pass else 1

if __name__ == "__main__":
    sys.exit(main())
